## BUSA8001- Programming Task 1  

**Assignment Points**: 100  
**Submission**: Provide your answers in this notebook and submit it via iLearn

- Where a question requires a written answer provide your solution in Markdown in the cells under each question.
- Comment out your print statements unless you are explicitly asked to print your output. 
- 5 marks will be deducted for printed outputs that are not asked for.

### About the Assignment

- For this assignment there are two files in the `data` folder `credit_record.csv` and `application_record.csv` where bank clients are related by the `ID` column.

- In `application_record.csv` we have the following variables

| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| AMT_INCOME   | Annual income  |  |
| NAME_INCOME_TYPE   | Income Source |  |
| NAME_EDUCATION_TYPE   | Level of Education  |  |
| CODE_GENDER   | Applicant's Gender   |  |
| FLAG_OWN_CAR | Car Ownership |  | 
| CNT_CHILDREN | Number of Children | |
| FLAG_OWN_REALTY | Real Estate Ownership | | 
| NAME_FAMILY_STATUS | Relationship Status | | 
| NAME_HOUSING_TYPE | Housing Type | | 
| DAYS_BIRTH | No. of Days | Count backwards from current day (0), -1 means yesterday
| DAYS_EMPLOYED | No. of Days | Count backwards from current day (0). If positive, it means the person is currently unemployed.
| FLAG_MOBIL | Mobile Phone Ownership | | 
| FLAG_WORK_PHONE | Work Phone Ownership | | 
| FLAG_PHONE | Landline Phone Ownership | | 
| FLAG_EMAIL | Landline Phone Ownership | | 
| OCCUPATION_TYPE | Occupation | | 
| CNT_FAM_MEMBERS | Count of Family Members | |



- In `credit_record.csv` we have the following variables


| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number | |
| MONTHS_BALANCE | Number of months in the past from now when STATUS is measured | 0 = current month, -1 = last month, -2 = two months ago, etc.|
| STATUS | Number of days a payment is past due | 0: 1-29 days past due 1: 30-59 days past due 2: 60-89 days overdue 3: 90-119 days overdue 4: 120-149 days overdue 5: Overdue or bad debts, write-offs for more than 150 days C: paid off that month X: No loan for the month |

---
---

### Task 1: Reading, Summarising and Cleaning Data (Total Marks: 30)

**Question 1.** 

1. Import the `application_record.csv` and `credit_record.csv` files from `data` folder into pandas DataFrames named `df_application` and `df_credit`, respectively. (1 mark)

2. How many rows are there in `df_application` and `df_credit`, respectively? Provide your answers with print() and state them in Markdown text. (1 mark)

3. How many unique bank clients are there in `df_application` and `df_credit`? Provide your answers with print() and state them in Markdown text. (1 mark)

4. Add the records from `df_credit` to `df_application` by merging the data from the two DataFrames on the `ID` column, and output the joint data into a new DataFrame named `df`. Hint: Use `merge` function from pandas by setting `how` parameter to `inner` (4 marks) 

5. How many rows and how many unique clients are there now in `df`? (1 mark)

6. How are multiple rows for each `ID` in `df` different? Answer in Markdown text. (2 mark) 

(10 marks)

In [1]:
# ---- provide your code here -----
# Question 1: 
# 1.Import the application_record.csv and credit_record.csv files from data folder into pandas DataFrames 
# named df_application and df_credit, respectively.

import pandas as pd
import numpy as np

# Load the CSV files
df_application = pd.read_csv('data/application_record.csv')
df_credit = pd.read_csv('data/credit_record.csv')

#df_application.head(10)
#df_credit.head(10)

In [2]:
# 2. How many rows are there in df_application and df_credit, respectively? 
# Provide your answers with print() and state them in Markdown text.

print("Number of rows in df_application:", df_application.shape[0])
print("Number of rows in df_credit:", df_credit.shape[0])


Number of rows in df_application: 438445
Number of rows in df_credit: 1047185


---
---
2.

Number of rows in df_application: 438445

Number of rows in df_credit:1047185

In [3]:
# 3. How many unique bank clients are there in df_application and df_credit? 
# Provide your answers with print() and state them in Markdown text.

print("Number of unique bank clients in df_application:", df_application['ID'].nunique())
print("Number of unique bank clients in df_credit:", df_credit['ID'].nunique())

Number of unique bank clients in df_application: 438398
Number of unique bank clients in df_credit: 45924


---
---
3.

Number of unique bank clients in df_application: 438398

Number of unique bank clients in df_credit:45924

In [4]:
# 4. Add the records from df_credit to df_application by merging the data from the two DataFrames on the ID column, 
# and output the joint data into a new DataFrame named df 
# Hint: Use merge function from pandas by setting how parameter to inner (4 marks)

df = pd.merge(df_application, df_credit, on='ID', how='inner')

#df.head(25)

In [5]:
# 5. How many rows and how many unique clients are there now in df?

print("Number of rows in df:", df.shape[0])
print("Number of unique clients in df:", df['ID'].nunique())

Number of rows in df: 776325
Number of unique clients in df: 36396


### 6. How are multiple rows for each ID in df different? Answer in Markdown text.

Although these IDs have same information like gender, income type, education type, etc., some IDs have different values in  "MONTH_BALANCE" AND 'STATUS' columns.

---- provide your written answer here ----


---

**Question 2.**

1. Change the values of `STATUS` in `df` according to the following mapping: {C, X, 0} -> 0 and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers. (2 marks)

2. Create a new *numpy* array called `list_of_past_due` that includes the unique ID numbers of clients whose `STATUS = 1` at any point during the last 12 months (hint: count the current month as the first month). (2 marks) 

3. Create a new DataFrame called `df_final` that contains the rows of `df` for which the `ID` is in `list_of_past_due`, keeping only one row for each `ID` (hint: keep the first duplicate row). How many rows do you have in `df_final`? Answer using both print() function and in Markdown text. (hint: find out about `isin()` function in pandas.) (2 marks)

4. Add a new column `y = 1` for all the rows in `df_final`. (1 marks)

5. Increase `df_final` to a total of 4,500 rows by adding rows from `df` with unique `ID`s which are not in `list_of_past_due`. To do this start adding the rows from the beginning of `df`. (hint: learn what `~`, i.e. tilde sign, does in pandas). (2 marks) 

6. Fill the missing values of `y` in `df_final` with zeros. Remove `STATUS` and `MONTHS_BALANCE` from `df_final`. (1 mark)

(10 marks)

In [6]:
# ---- provide your code here -----
# Question 2:
# 1. Change the values of STATUS in df according to the following mapping: {C, X, 0} -> 0 
# and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers.

status_mapping = {'C': 0, 'X': 0, '0': 0, '1': 1, '2': 1, '3': 1, '4': 1, '5': 1}
df['STATUS'] = df['STATUS'].map(status_mapping).astype(int)

In [7]:
# 2. Create a new numpy array called list_of_past_due 
# that includes the unique ID numbers of clients whose STATUS = 1 at any point during the last 12 months 
# (hint: count the current month as the first month).

recent_12_months = df[df['MONTHS_BALANCE'] >= -11]
list_of_past_due = np.array(recent_12_months[recent_12_months['STATUS'] == 1]['ID'].unique())

In [8]:
print(len(list_of_past_due))

1737


In [9]:
# 3. Create a new DataFrame called df_final that contains the rows of df for which the ID is in list_of_past_due, 
# keeping only one row for each ID (hint: keep the first duplicate row). 
# How many rows do you have in df_final? Answer using both print() function and in Markdown text. (hint: find out about isin() function in pandas.) 

df_final = df[df['ID'].isin(list_of_past_due)].drop_duplicates(subset=['ID'], keep='first')

In [10]:
print("Number of rows in df_final:", df_final.shape[0])

Number of rows in df_final: 1737


---
Number of rows in df_final: 1737

In [11]:
# 4. Add a new column y = 1 for all the rows in df_final. 

df_final['y'] = 1
#df_final.head()

In [12]:
# 5. Increase df_final to a total of 4,500 rows by adding rows from df with unique IDs which are not in list_of_past_due. 
# To do this start adding the rows from the beginning of df. (hint: learn what ~, i.e. tilde sign, does in pandas).

remain_id = df[~df['ID'].isin(list_of_past_due)]['ID'].unique()
add_rows = df[df['ID'].isin(remain_id)].drop_duplicates(subset=['ID'], keep='first')

df_final = pd.concat([df_final, add_rows]).head(4500)
#df_final.head(1800)

In [13]:
# 6. Fill the missing values of y in df_final with zeros. Remove STATUS and MONTHS_BALANCE from df_final.

df_final['y'] = df_final['y'].fillna(0).astype(int)
df_final = df_final.drop(columns=['STATUS', 'MONTHS_BALANCE'])

---- provide your written answer here ----


<hr style="width:25%;margin-left:0;"> 

**Question 3**. 
1. Delete `ID` column from `df_final` and reset its index. (1 marks)
2. Assuming that `NAME_EDUCATION_TYPE` is the only ordinal variable in `df_final`, which variables are numeric and which ones are nominal? Answer this question by copying and completing the following table (6 marks)

|Variable type|Numbers of features|Features' list|
| --- | --- | --- |
|Numeric:|||
|Ordinal:|1| NAME_EDUCATION_TYPE |
|Nominal:|||

3. Using appropriate functions find and comment on the missing values in `df_final` (3 marks)   

(10 marks)

In [14]:
# ---- provide your code here ----
# Question 3: 
# 1. Delete ID column from df_final and reset its index. 

df_final = df_final.drop(columns=['ID']).reset_index(drop=True)

In [15]:
# Assuming that NAME_EDUCATION_TYPE is the only ordinal variable in df_final, 
# which variables are numeric and which ones are nominal? 
# Answer this question by copying and completing the following table
df_final.head()

CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME  \
0           F            N               Y           NaN    157500.0   
1           M            Y               Y           NaN    360000.0   
2           F            N               Y           0.0    297000.0   
3           F            N               Y           0.0    297000.0   
4           F            N               Y           0.0    297000.0   

       NAME_INCOME_TYPE            NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  \
0               Working  Secondary / secondary special               Married   
1  Commercial associate                            NaN               Married   
2  Commercial associate                            NaN  Single / not married   
3  Commercial associate                            NaN  Single / not married   
4  Commercial associate                            NaN  Single / not married   

   NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  FLAG_WORK_PHONE  \
0  House / apartment      -10031          -1469           1                0   
1  House / apartment      -16670          -5364           1                0   
2   Rented apartment      -15519          -3234           1                0   
3   Rented apartment      -15519          -3234           1                0   
4   Rented apartment      -15519          -3234           1                0   

   FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  y  
0           1           0        Laborers                2  1  
1           1           0  Security staff                2  1  
2           0           0        Laborers                1  1  
3           0           0        Laborers                1  1  
4           0           0        Laborers                1  1

|Variable type|Numbers of features|Features' list|
| --- | --- | --- |
|Numeric:|10| CNT_CHILDREN, AMT_INCOME, DAYS_BIRTH, DAYS_EMPLOYED, FLAG_MOBIL, FLAG_WORK_PHONE, FLAG_PHONE, FLAG_EMAIL, CNT_FAM_MEMBERS, y |
|Ordinal:|1| NAME_EDUCATION_TYPE |
|Nominal:|7| CODE_GENDER, FLAG_OWN_CAR, FLAG_OWN_REALTY, NAME_INCOME_TYPE, NAME_FAMILY_STATUS, NAME_HOUSING_TYPE, OCCUPATION_TYPE |

In [16]:
# 3. Using appropriate functions find and comment on the missing values in df_final 

missing_values = df_final.isnull().sum()
print(missing_values)

CODE_GENDER               0
FLAG_OWN_CAR              0
FLAG_OWN_REALTY           0
CNT_CHILDREN             74
AMT_INCOME                0
NAME_INCOME_TYPE          0
NAME_EDUCATION_TYPE    1831
NAME_FAMILY_STATUS        0
NAME_HOUSING_TYPE         0
DAYS_BIRTH                0
DAYS_EMPLOYED             0
FLAG_MOBIL                0
FLAG_WORK_PHONE           0
FLAG_PHONE                0
FLAG_EMAIL                0
OCCUPATION_TYPE        1354
CNT_FAM_MEMBERS           0
y                         0
dtype: int64


---- provide your written answer here ----


---
---

### Task 2: Imputing missing values and dealing with categorical features (Total Marks: 30)



**Question 4.** 
1. Use an appropriate `pandas` function to impute missing values in `df_final` (15 marks)
    - Take into consideration the type of each variable and the best practices we discussed in class/lecture notes

(Total: 15 marks)

In [17]:
# ---- provide your code here -----

#Question 4.

#Use an appropriate pandas function to impute missing values in df_final (15 marks)
#Take into consideration the type of each variable and the best practices we discussed in class/lecture notes

# Impute numeric variable CNT_CHILDREN with median
df_final['CNT_CHILDREN'].fillna(df_final['CNT_CHILDREN'].median(), inplace=True)

# Impute ordinal variable NAME_EDUCATION_TYPE with mode (most frequent value)
df_final['NAME_EDUCATION_TYPE'].fillna(df_final['NAME_EDUCATION_TYPE'].mode()[0], inplace=True)

# Impute nominal variable OCCUPATION_TYPE with 'NA'
df_final['OCCUPATION_TYPE'].fillna('NA', inplace=True)


<hr style="width:25%;margin-left:0;"> 

**Question 5**. Convert the values in `NAME_EDUCATION_TYPE` as follows
- Lower secondary -> 1
- Secondary / secondary special -> 2
- Incomplete higher -> 3
- Higher education -> 4


(Total: 5 marks)  

In [18]:
# ---- provide your code here -----
#Question 5. Convert the values in NAME_EDUCATION_TYPE as follows

#Lower secondary -> 1
#Secondary / secondary special -> 2
#Incomplete higher -> 3
#Higher education -> 4

from sklearn.preprocessing import OrdinalEncoder
ordinal_encoder = OrdinalEncoder(categories=[['Lower secondary', 
                                              'Secondary / secondary special', 
                                              'Incomplete higher', 
                                              'Higher education']])
# Reshape the data to use with OrdinalEncoder
df_final['NAME_EDUCATION_TYPE'] = ordinal_encoder.fit_transform(df_final[['NAME_EDUCATION_TYPE']])

# Increment values to match your desired encoding (1 to 4 instead of 0 to 3)
df_final['NAME_EDUCATION_TYPE'] = df_final['NAME_EDUCATION_TYPE'].astype(int) + 1

<hr style="width:25%;margin-left:0;"> 

**Question 6**. 

Add dummy variables to `df_final` for all of the nominal features which are currently stored as string (text). 
- Make sure to delete the original variables from the dataframe
- Drop the first column from each set of created dummy variable, i.e. for each feature



(Total: 10 marks)  

In [19]:
# ---- provide your code here -
#Question 6.

#Add dummy variables to df_final for all of the nominal features which are currently stored as string (text).

#Make sure to delete the original variables from the dataframe
#Drop the first column from each set of created dummy variable, i.e. for each feature

nominal_columns = ['CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 
                   'NAME_INCOME_TYPE', 'NAME_FAMILY_STATUS', 
                   'NAME_HOUSING_TYPE', 'OCCUPATION_TYPE']

# Create dummy variables and drop the original columns
df_final = pd.get_dummies(df_final, columns=nominal_columns, drop_first=True)

In [20]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4500 entries, 0 to 4499
Data columns (total 45 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   CNT_CHILDREN                             4500 non-null   float64
 1   AMT_INCOME                               4500 non-null   float64
 2   NAME_EDUCATION_TYPE                      4500 non-null   int32  
 3   DAYS_BIRTH                               4500 non-null   int64  
 4   DAYS_EMPLOYED                            4500 non-null   int64  
 5   FLAG_MOBIL                               4500 non-null   int64  
 6   FLAG_WORK_PHONE                          4500 non-null   int64  
 7   FLAG_PHONE                               4500 non-null   int64  
 8   FLAG_EMAIL                               4500 non-null   int64  
 9   CNT_FAM_MEMBERS                          4500 non-null   int64  
 10  y                                        4500 no

---
---

### Task 3 Preparing X and y arrays (Total Marks: 10)

**Question 7**. 

1. Create a numpy array named `y` from the `y` column of `df_final` making sure that the values of the array `y` are stored as integers (3 marks)   
2. Create a numpy array named `X`  from all the remaining features in `df_final` (2 marks)   

(Total: 5 Marks)

In [21]:
# ---- provide your code here -----
#Question 7.

#Create a numpy array named y from the y column of df_final making sure that the values of the array y are stored as integers (3 marks)
#Create a numpy array named X from all the remaining features in df_final (2 marks)

# Create a numpy array 'y' from the 'y' column and ensure it's stored as integers
y = np.array(df_final['y'].astype(int))

# Create a numpy array 'X' from all the remaining features
# Drop the 'y' column from df_final and convert the remaining columns to a NumPy array using np.array
X = np.array(df_final.drop(columns=['y']))

# Check the shapes of X and y
print("Shape of X:", X.shape)
print("Shape of y:", y.shape)


Shape of X: (4500, 44)
Shape of y: (4500,)


<hr style="width:25%;margin-left:0;"> 

**Question 8**. 

1. Use an appropriate scikit-learn library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 75% train and 25% test datasets (2.5 marks) 
    - Set random_state to 8 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
2. Standardise the data using `StandardScaler` library (2.5 marks)   

(Total: 5 marks) 

In [22]:
# ---- provide your code here -----
# Question 8.

#Use an appropriate scikit-learn library we used in class to create y_train, y_test, X_train and X_test by splitting the data into 75% train and 25% test datasets (2.5 marks)
#Set random_state to 8 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels
#Standardise the data using StandardScaler library (2.5 marks)

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=8, stratify=y)\

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

scaler = StandardScaler()

# Fit the scaler to the training data and transform it
X_train = scaler.fit_transform(X_train)

# Apply the same transformation to the test data
X_test = scaler.transform(X_test)

#print("First few rows of standardized X_train:", X_train[:5])

Shape of X_train: (3375, 44)
Shape of X_test: (1125, 44)
Shape of y_train: (3375,)
Shape of y_test: (1125,)


---
---

### Task 4. Logistic Regression and Random Forest Classifiers and Accuracies (Total Marks: 30)


**Question 9**. 

1. Train a Logistic Regression Classifier on standardised data (5 marks)
    - Set `random_state` to 10 (don't change any other parameters)
    - Compute and print training and test dataset accuracies   
2. Train a Random Forest Classifier on standardised data (5 marks)
    - Set `random_state` to 10 (don't change any other parameters)
    - Compute and print training and test dataset accuracies

When printing accuracies round the values to three decimal places.      

(Total: 10 marks)  

In [23]:
# ---- provide your code here -----

#Question 9.

# 1. Train a Logistic Regression Classifier on standardised data (5 marks)
#Set random_state to 10 (don't change any other parameters)
#Compute and print training and test dataset accuracies

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initialize the Logistic Regression classifier with a fixed random state
log_reg = LogisticRegression(random_state=10)

# Train the classifier on the standardized training data
log_reg.fit(X_train, y_train)

# Predict on the training and test data
y_train_pred = log_reg.predict(X_train)
y_test_pred = log_reg.predict(X_test)

# Calculate training and test accuracies
accuracy_train_log_reg = accuracy_score(y_train, y_train_pred)
accuracy_test_log_reg = accuracy_score(y_test, y_test_pred)

# Print the accuracies rounded to three decimal places
print("Logistic Regression Training Accuracy:", round(accuracy_train_log_reg, 3))
print("Logistic Regression Test Accuracy:", round(accuracy_test_log_reg, 3))

Logistic Regression Training Accuracy: 0.657
Logistic Regression Test Accuracy: 0.654


In [24]:
#3. Train a Random Forest Classifier on standardised data (5 marks)
#Set random_state to 10 (don't change any other parameters)
#Compute and print training and test dataset accuracies

from sklearn.ensemble import RandomForestClassifier

# Initialize the Random Forest classifier with a fixed random state
random_forest = RandomForestClassifier(random_state=10)

# Train the classifier on the standardized training data
random_forest.fit(X_train, y_train)

# Predict on the training and test data
y_train_pred_rf = random_forest.predict(X_train)
y_test_pred_rf = random_forest.predict(X_test)

# Calculate training and test accuracies
accuracy_train_rf = accuracy_score(y_train, y_train_pred_rf)
accuracy_test_rf = accuracy_score(y_test, y_test_pred_rf)

# Print the accuracies rounded to three decimal places
print("Random Forest Training Accuracy:", round(accuracy_train_rf, 3))
print("Random Forest Test Accuracy:", round(accuracy_test_rf, 3))

Random Forest Training Accuracy: 0.977
Random Forest Test Accuracy: 0.897


**Question 10**. 

a) Comment and compare the training and test accuracies for each classifier computed in Question 9. What can we say about the extent of overfitting for each classifier? (5 marks)   
b) Comment and compare the accuracies across the two classifiers. Which classifier provides better forecasts? (5 marks)   
c) What can you say about the presence of nonlinearities in the dataset? (10 marks)   

(Total: 20 marks)  

---
---

### Part a

Logistic Regression:

Training Accuracy: 0.657
Test Accuracy: 0.654
The training and test accuracies for the Logistic Regression model are quite close, indicating minimal overfitting. The model generalizes well to unseen data (test set). However, the overall accuracy is moderately low, suggesting that the model may be underfitting the dataset or that the linear model may not be capturing all the complexities of the data.

Random Forest:

Training Accuracy: 0.977
Test Accuracy: 0.897
The Random Forest classifier shows a high training accuracy but a lower test accuracy. The discrepancy between these two metrics suggests some degree of overfitting. The model performs exceptionally well on the training data, likely capturing intricate patterns that do not generalize as effectively on the test data. However, even with this overfitting, the test accuracy is significantly higher than that of the Logistic Regression model.

---
---

### Part b

Random Forest provides better forecasts than Logistic Regression, as evidenced by its higher test accuracy (0.897 vs. 0.654). This indicates that the Random Forest model is more capable of handling the complexity and variability in the data.
Logistic Regression, while not overfitting, shows lower accuracy both in training and testing, which might indicate it is too simplistic a model to capture the necessary patterns in the data for more accurate predictions.

---
---

### Part c

The performance differences between the Logistic Regression and Random Forest classifiers can provide insights into the nature of the dataset:

Nonlinear Relationships: The fact that the Random Forest, which can handle complex nonlinear relationships between features, significantly outperforms Logistic Regression suggests the presence of nonlinearities in the data. Logistic Regression's assumption of linearity appears to be insufficient for modeling the underlying relationships effectively.
Model Suitability: The superior performance of Random Forest indicates that the dataset benefits from a model that can capture interactions and nonlinearities. This is typical in real-world scenarios where many factors interact in complex ways that a linear model like Logistic Regression cannot adequately capture.
Data Complexity: The high training accuracy of Random Forest also implies that there are likely meaningful patterns in the data, which this model can learn but may also overfit. This suggests that careful feature selection, parameter tuning, or the application of regularization techniques might be necessary to optimize model performance further without losing the ability to generalize.